### PR TITLE
Custom metrics and callbacks complete episodes

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This is an implementation of two suggestions made by @sven1977 in the PR #22684 . 

## Related issue number

None

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
the commit history does look kinda crazy. what I usually do is:
```git pull --rebase``` in master, then checkout the feature branch, then```git rebase master```, then ```git push -f ...```.
that way it's always based off master HEAD.

File: rllib/evaluation/collectors/sample_collector.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def __init__(self,
                  policy_map: PolicyMap,
                  clip_rewards: Union[bool, float],
                  callbacks: "DefaultCallbacks",
-                 multiple_episodes_in_batch: bool = True,
+                 collect_complete_episodes: bool = False,
```
Comment (gjoliver) [inline rllib/evaluation/collectors/sample_collector.py:35]: can we not change the default, this may be a huge change.

**Review by gjoliver — COMMENTED** [review]
concerned about changing all these defaults. that will impact tons of users with they even realize.

**Comment by simonsays1980** [comment]
> concerned about changing all these defaults. that will impact tons of users with they even realize.

@gjoliver Thanks for your valuable input! This is a fair point, I haven't thought about this when I implemented the points @sven1977 made in the linked PR. 

### Suggestion
What do you think about the following: 

- Whenever users have set `multiple_episodes_in_batch` in any function, the new implementation would still receive these settings via the `**kwargs` parameter
- We could then simply check for `multiple_episodes_in_batch` in the `**kwargs` and set `collect_complete_episodes` appropriately
- In addition we then mark the usage of `multiple_episodes_in_batch` as `deprecated` to introduce the new way 
This could ensure a smooth transition from using `multiple_episodes_in_batch` to `collect_complete_episodes` with backward compatability.

Is this a way how you guys usually introduce changes in parameters? I am looking forward to your feedback.

**Comment by simonsays1980** [comment]
> the commit history does look kinda crazy. what I usually do is: `git pull --rebase` in master, then checkout the feature branch, then`git rebase master`, then `git push -f ...`. that way it's always based off master HEAD.

Yup, apologies! I did not recognize these changes when I rebased in my branch. I will follow your suggestion in the future of this branch and others. Thanks for the tip!

**Comment by gjoliver** [comment]
> > concerned about changing all these defaults. that will impact tons of users with they even realize.
> 
> @gjoliver Thanks for your valuable input! This is a fair point, I haven't thought about this when I implemented the points @sven1977 made in the linked PR.
> 
> ### Suggestion
> What do you think about the following:
> 
> * Whenever users have set `multiple_episodes_in_batch` in any function, the new implementation would still receive these settings via the `**kwargs` parameter
> * We could then simply check for `multiple_episodes_in_batch` in the `**kwargs` and set `collect_complete_episodes` appropriately
> * In addition we then mark the usage of `multiple_episodes_in_batch` as `deprecated` to introduce the new way
>   This could ensure a smooth transition from using `multiple_episodes_in_batch` to `collect_complete_episodes` with backward compatability.
> 
> Is this a way how you guys usually introduce changes in parameters? I am looking forward to your feedback.

@sven1977 

> > concerned about changing all these defaults. that will impact tons of users with they even realize.
> 
> @gjoliver Thanks for your valuable input! This is a fair point, I haven't thought about this when I implemented the points @sven1977 made in the linked PR.
> 
> ### Suggestion
> What do you think about the following:
> 
> * Whenever users have set `multiple_episodes_in_batch` in any function, the new implementation would still receive these settings via the `**kwargs` parameter
> * We could then simply check for `multiple_episodes_in_batch` in the `**kwargs` and set `collect_complete_episodes` appropriately
> * In addition we then mark the usage of `multiple_episodes_in_batch` as `deprecated` to introduce the new way
>   This could ensure a smooth transition from using `multiple_episodes_in_batch` to `collect_complete_episodes` with backward compatability.
> 
> Is this a way how you guys usually introduce changes in parameters? I am looking forward to your feedback.

yeah, I don't know if it's worth it to go through all the deprecation pain. @sven1977 
since you are checking ```worker.policy_config["batch_mode"]``` now in your callback handling, this argument change is not really needed right?

**Pull request closed without merging**